### PR TITLE
[ll] D3D12 backend MVP

### LINKED ISSUE #1353: [ll] Mapping size/offset semantics
Currently, `write_mapping` gets the `offset` and `size` parameters in `T` units. This appears slightly inconsistent with other `Factory` methods (e.g. `create_buffer`) that work with bytes. I believe that even though we technically can treat the buffer as typed, given the `T` knowledge, we should be consistent with our `offset` and `size`, and treat them in bytes.

@msiglreith thoughts?

### PR DESCRIPTION
**Pull request by kvark**
Fixes #1353
Also brings some improvements to core and metal backends.

### CODE REVIEW

File: src/backend/dx12ll/src/factory.rs [hunk 1]
```diff
@@ -97,17 +97,71 @@ impl Factory {
         }
         Ok(native::ShaderLib { shaders: shader_map })
     }
+
+    pub fn create_descriptor_heap_impl(device: &mut ComPtr<winapi::ID3D12Device>,
```
Comment (msiglreith) [inline src/backend/dx12ll/src/factory.rs]: Could you annotate that this as an non-shader visible/cpu heap?

File: src/backend/dx12ll/src/lib.rs [hunk 2]
```diff
@@ -130,8 +136,26 @@ impl core::Adapter for Adapter {
...
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/dn788678(v=vs.85).aspx
+        let heap_types = vec![
+            core::HeapType {
+                id: 0,
+                properties: memory::DEVICE_LOCAL,
+                heap_index: 1,
+            },
+            core::HeapType {
+                id: 1,
+                properties: memory::CPU_VISIBLE | memory::CPU_CACHED,
```
Comment (msiglreith) [inline src/backend/dx12ll/src/lib.rs]: Does D3D12 have non-coherent memory? Vulkan requires to explicitly flush non-coherent memory via `vkFlush/InvalidateMappedMemoryRanges`. I don't recall if there is a similar restriction on d3d12 (map/unmap?).

**Comment by kvark** [comment]
Progress is real, landing imminent:
gfx-dx12

File: src/backend/dx12ll/src/command.rs [hunk 3]
```diff
@@ -1,575 +1,726 @@
...
+    }
+
+    fn copy_image_to_buffer(&mut self) {
+        unimplemented!()
+    }
+
+    fn clear_color(&mut self, rtv: &native::RenderTargetView, value: command::ClearColor) {
+        let clear_color = match value {
+            command::ClearColor::Float(c) => c,
+            command::ClearColor::Int(c) => [c[0] as FLOAT, c[1] as FLOAT, c[2] as FLOAT, c[3] as FLOAT], // TODO: error?
```
Comment (msiglreith) [inline src/backend/dx12ll/src/command.rs]: (Casts are probably wrong here, requiring to keep the same bits but that's something we rather need to test in the future..)

File: src/backend/dx12ll/src/command.rs [hunk 4]
```diff
@@ -1,575 +1,726 @@
...
+macro_rules! impl_render_pass_inline_buffer {
+    ($graphics_buffer:ident) => {
+        impl command::RenderPassInlineBuffer<$graphics_buffer, R> for RenderPassInlineBuffer {
+            fn begin(command_buffer: &mut Encoder<$graphics_buffer>,
+                     render_pass: &RenderPass,
+                     framebuffer: &FrameBuffer,
+                     render_area: target::Rect,
+                     clear_values: &[command::ClearValue]
+            ) -> Self {
+				let color_views = framebuffer.color.iter().map(|view| view.handle).collect::<Vec<_>>();
```
Comment (msiglreith) [inline src/backend/dx12ll/src/command.rs]: nit: formatting

File: src/backend/dx12ll/src/command.rs [hunk 5]
```diff
@@ -1,575 +1,726 @@
...
+       			assert!(framebuffer.depth_stencil.len() <= 1);
```
Comment (msiglreith) [inline src/backend/dx12ll/src/command.rs]: Whats the motivation behind this check?

File: src/backend/dx12ll/src/native.rs [hunk 6]
```diff
@@ -115,28 +125,92 @@ pub struct DepthStencilView {
...
+#[derive(Clone, Debug)]
+pub struct DualHandle {
+    pub cpu: winapi::D3D12_CPU_DESCRIPTOR_HANDLE,
+    pub gpu: winapi::D3D12_GPU_DESCRIPTOR_HANDLE,
+}
+
+#[derive(Clone, Debug)]
 pub struct DescriptorHeap {
+    pub inner: ComPtr<winapi::ID3D12DescriptorHeap>,
+    //TODO: hide members?
```
Comment (msiglreith) [inline src/backend/dx12ll/src/native.rs]: Sounds like a good place for `pub(crate)` 😄

**Comment by kvark** [comment]
Looks green enough :) Merging, at last!

**Pull request merged**